### DISCUSSION TITLE
I need help with SlicingPlane.

### DISCUSSION THREAD

**Discussion by JHsGame** [mlavik1/UnityVolumeRendering#86] [Q&A]
Hello, it's a good day. I would appreciate it if you could understand that I don't speak English and use a translator.

It is intended to be available among plays other than SlicingPlane editors.

So I'm going to get the code that the producer has implemented and click the button to add Slicing Plane.

But the problem here is...

sliceMat.SetTexture("_TFTex", transferFunction.GetTexture());

In the part, a value should be included in the _TFTex, but I'm not sure where to get the transferFunction.

The difference between the plane taken from editor mode and the plane made in play mode was whether the_TFTex was null or not.

```
public void OnClickCreateSlicingPlane()
{
       CreateSlicingPlane();
}

public SlicingPlane CreateSlicingPlane()
    {
        transferFunction.GenerateTexture();
        transferFunction2D.GenerateTexture();

        GameObject sliceRenderingPlane = GameObject.Instantiate(Resources.Load<GameObject>("SlicingPlane"));
        sliceRenderingPlane.transform.parent = myVolume.transform.parent;
        sliceRenderingPlane.transform.localPosition = Vector3.zero;
        sliceRenderingPlane.transform.localRotation = Quaternion.identity;
        sliceRenderingPlane.transform.localScale = Vector3.one * 0.1f; // TODO: Change the plane mesh instead and use Vector3.one
        MeshRenderer sliceMeshRend = sliceRenderingPlane.GetComponent<MeshRenderer>();
        sliceMeshRend.material = new Material(sliceMeshRend.sharedMaterial);
        Material sliceMat = sliceRenderingPlane.GetComponent<MeshRenderer>().sharedMaterial;
        sliceMat.SetTexture("_DataTex", texture);
        sliceMat.SetTexture("_TFTex", transferFunction.GetTexture());
        sliceMat.SetMatrix("_parentInverseMat", myVolume.transform.parent.worldToLocalMatrix);
        sliceMat.SetMatrix("_planeMat", Matrix4x4.TRS(sliceRenderingPlane.transform.position, sliceRenderingPlane.transform.rotation, myVolume.transform.parent.lossyScale)); // TODO: allow changing scale


        GameObject texturePlane = GameObject.Instantiate(Resources.Load<GameObject>("TexturePlane"));
        texturePlane.transform.parent = TexturePlaneParent.transform;
        texturePlane.transform.localPosition = new Vector3(texturePlane.transform.position.x * TexturePlaneParent.transform.childCount, texturePlane.transform.position.y, texturePlane.transform.position.z);
        texturePlane.transform.localRotation = Quaternion.Euler(-90f, 0f, 0f);
        texturePlane.GetComponent<TestTexture>().obj = sliceRenderingPlane;


        return sliceRenderingPlane.GetComponent<SlicingPlane>();
    }
```
When you click the button in play mode, you can create a SlicingPlane by executing the CreateSlicingPlane() function with the OnClickCreateSlicingPlane() function.

**Comment by mlavik1** [comment]
Hi! I'm not 100% sure if I understand your questions, but is this the code in VolumeRenderedObject.cs?

_"... , but I'm not sure where to get the transferFunction"_
-> The transfer function is stored in the `VolumeRenderedObject`, which is a component that is attached to the GameObject of the imported dataset. You can get all VolumeRenderedObject instances in the scene like this: `VolumeRenderedObject[] spawnedObjects = GameObject.FindObjectsOfType<VolumeRenderedObject>();`

_The difference between the plane taken from editor mode and the plane made in play mode was whether the_TFTex was null or not._
-> Hmm, I was not able to reproduce that. Exactly how can I test it? I tried importing a dataset, then enter play mode and then add a SlicingPlane from the menu => that worked fine here.

Have a nice weekend! :)

**Reply by JHsGame** [reply]
As you said, I brought the code on VolumeRenderedObject.cs and modified some parts.

It doesn't seem to have been delivered properly using a translator, but in order to create a Slice Render, you have to select it from the Volume Rendering tab of the editor.
But what I want is to make a separate button so that I can do it even after I build it.

That's why I brought the code and added it, and it succeeded in creating a plane containing the Slice Render by clicking the button, but as I said, the value of transferFunction.GetTexture() is NULL because it does not contain the value of transferFunction. As a result, even if you move the plane on an object that includes the Volume Render, the desired shape does not come out and it comes out white like the attached image.
(TransferFunction is not defined, resulting in errors, so all of these phrases are annotated.)

I think that as long as the value of transferFunction goes in, a proper output will come out. I would appreciate it if you could help me modify the code so that I can use it even after building it as well as the editor.
![1](https://user-images.githubusercontent.com/80305896/158083352-d9f4717c-4cbb-4b81-b83d-7bcb63422162.png)
![2](https://user-images.githubusercontent.com/80305896/158083353-7048404d-722c-426a-b4c4-778d121701a1.png)

**Comment by mlavik1** [comment]
Hmm, and the `CreateSlicingPlane` function is still inside VolumeRenderedObject.cs, right?
Because you need to use the transfer function of the dataset / VolumeRenderedObject.
In your screenshot I can see that the volume has a valid transfer function, and then transferFunction.GetTexture() should not return Null. So I have a feeling that maybe you are trying to use the wrong transfer function? It should be the one from `VolumeRenderedObject.transferFunction`

Is your source code available somewhere, so I can see it? It's difficult to see what's wrong without having the code available :sweat_smile:

**Reply by JHsGame** [reply]
Uh... I think it's hard to understand because I'm customizing your Asset separately so it can be used in play mode.

Therefore, I will send you a separate e-mail of Unity that I am currently touching.

If you are using TestScene in the Scenes folder and are in play mode, you can simply load the folder containing the .dcm folder into the file input/output. Then click the Test Create SlicingPlane button at the bottom of the screen to create the SlicingPlane I mentioned. You can see that this is created in Hierachy.

The scripts related to this are FileLoader and TestController in the Script folder.

For your information, you have to add the package at the following address to Unity so that the error does not appear.

https://github.com/karl-/pb_Stl

I really hope the file I sent you solves my problem.

I'm really sorry to bother you, and I hope you have a good day. :)

**Reply by mlavik1** [reply]
Hi! I just replied to you on e-mail :)

**Reply by leibm** [reply]
Wow, I also encountered a similar problem to you, have you solved it?
